### PR TITLE
feat: repository cloner

### PR DESCRIPTION
**Pull request by sgaist**
## Describe your changes

This PR implements support for the CodeRepository management.

The added functionalities:
- Cloner helper program that supports both username/password and ssh key based authentication
- Creation of init containers to clone the repositories.

The current implementation creates one init container per CodeRepository entry.

**Known limitations**:
- While the `cloner` helper supports ssh remotes, it will fail when run in a container as the remote will not be part of the known hosts. One solution would be to provide the known_hosts file through a ConfigMap mounted in the init containers. This would allow users to add private registries in a secure fashion.
- When restarting the pod, clone will fail as the target already exists.

For that last point two possible things can be done:
- Test if the folder exists and if so don't clone and call it a success
- Add an annotation to the pod that can be checked and not re-create the clone containers.

Tests are currently a skeleton and must be improved.

## Issue ticket number and link

Fixes: #612
Part of https://github.com/SwissDataScienceCenter/renku/issues/3679

### CODE REVIEW

**Review by olevski — CHANGES_REQUESTED** [review]
Thanks @sgaist, this looks really good.

A few things:
- Please check if cloning with username and password saves the credentials in the repository's git configuration in plain text. If I am not mistaken this is what happens when you clone with the regular git cli and use credentials. If the credentials are stored you can modify the repo's git config with functions from the `go-git` package.
- I think we should rename `AuthConfig` to `Config` or `CloneConfig` because I think we will need to add more stuff there in the future. For example we may want to support cloning via a proxy. 
- Can you please check if the `PlainClone` function actually respects/loads the global/user git configs at all. My impression is that it does not. I looked through the code for it and it seems like there are no calls to load the user or global git configs. It just loads the config inside the repository after the repository is initialized. If the global/user configs are not loaded at all then we should change the CRD spec and simply not ask for the git config to be used when cloning.

If you want just open issues for all the three things above here. And we can tackle them after this is merged.

**Comment by olevski** [comment]
One more thing... We can actually clone a repository by using the git config. You just have to use the lower level function from go-git. See here for example: https://gist.github.com/olevski/606c6168b6072cabadaca09e89744414

File: cloner/cmd/clone.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+		clonePath = path + "/" + projectName
+	}
+
+	// Clone the given repository to the given directory
+	log.Print("git clone ", remote, " to ", clonePath)
+
+	cloneOptions := git.CloneOptions{
+		URL:               remote,
+		SingleBranch:      true,
+		ReferenceName:     plumbing.ReferenceName(revision),
```
[the remark this answers is not in the document]
Comment (sgaist) [reply cloner/cmd/clone.go:101]: If ReferenceName is invalid, the clone will fail with a good error message so I don't think we should add our own.

**Comment by sgaist** [comment]
> Thanks @sgaist, this looks really good.
> 
> A few things:
> 
>     * Please check if cloning with username and password saves the credentials in the repository's git configuration in plain text. If I am not mistaken this is what happens when you clone with the regular git cli and use credentials. If the credentials are stored you can modify the repo's git config with functions from the `go-git` package.

From tests, the auth information are not stored in the cloned repository.

>     * I think we should rename `AuthConfig` to `Config` or `CloneConfig` because I think we will need to add more stuff there in the future. For example we may want to support cloning via a proxy.
> 

Good point and done

>     * Can you please check if the `PlainClone` function actually respects/loads the global/user git configs at all. My impression is that it does not. I looked through the code for it and it seems like there are no calls to load the user or global git configs. It just loads the config inside the repository after the repository is initialized. If the global/user configs are not loaded at all then we should change the CRD spec and simply not ask for the git config to be used when cloning.
> 

You are correct. There's a long talk about it here: https://github.com/src-d/go-git/issues/760

**Review by olevski — APPROVED** [review]

**Pull request merged**